### PR TITLE
Lido app/marcus/xlookup

### PR DESCRIPTION
**Pull request by marhong2**
Draft for @tyleradams, All questions are labeled as @TYLER . The main issues I am having is what should Filter return (RN it is an array of arrays) and if that is correct, how do I test it?

Thank you! 
Marcus

### CODE REVIEW

File: formulas/helpers.js [hunk 1]
```diff
@@ -299,6 +299,73 @@ class FormulaHelpers {
...
+     * ? any character can go here, For example, sm?th finds "smith" and "smyth"
+     * * Any number of characters, For example, *east finds "Northeast" and "Southeast"
+     * ~ (tilde) followed by ?, *, or ~ A question mark, asterisk, or tilde, For example, fy06~? finds "fy06?"
+     ***/
+     XLOOKUP_HELPER(lookup_value, compare_value, match_mode){
+        //in the event we are looking for the next largest/smallest value
+        compare_value = compare_value.toString();
+        var sum = 0;
+        if(match_mode != 2){
+            var min_Index = Math.min(lookup_value.length, compare_value.length);
```
[the remark this answers is not in the document]
Comment (marhong2) [reply formulas/helpers.js:325]: I rewrote the documentation, it should also give the magnitude of difference. This way I can implement the comparison based match modes. E.g. if I want to compare a and b, not only should it be a negative number, but should also return a number of smaller magnitude than a compared to z. This allows me to do comparisons for search mode 1 and -1! I see that the magnitude is not needed for BS so I'll implement your changes locally in XLOOKUP

[the remark this answers is not in the document]
Comment (marhong2) [reply formulas/helpers.js:325]: I think, locale compare only returns a value based upon their difference, not including their magnitude...

File: test/formulas/reference/testcase.js [hunk 2]
```diff
@@ -134,6 +150,233 @@ module.exports = {
         'VLOOKUP("a*", {"c";TRUE;"AbC";65;63;61;"b";"a";TRUE;FALSE}, 1, FALSE)': 'AbC',
         // single row table
         'VLOOKUP(614, { 614,"Foobar"}, 2)': 'Foobar'
+    },
+    XLOOKUP: {
+        //(0,1) (1,1) (-1,1) (2,1) (0,-1) (-1,-1) (1,-1) (2,-1) (0,2)
+        //'XLOOKUP(A2, A1:A9, B1:B9)': .69,
+        //match_mode = 0 and searc_mode is 1
```
Comment (tyleradams) [inline test/formulas/reference/testcase.js:157]: Good comments! Fix typos

**Review by tyleradams — CHANGES_REQUESTED** [review]
Overall this is more complicated than we'd like which makes it hard to validate correctness.

However, if you can show screenshots where excel's xlookup returns the same values as this, it's fine by me.

**Comment by marhong2** [comment]
I have double checked all formulas against excel up as I have been working!

**Comment by tyleradams** [comment]
Excellent!

**Comment by tyleradams** [comment]
> Would you like me to keep a catalog from here on out?
Just sign off that you verified them all against excel at the end.

> Also, excel does a weird thing where it doesn't think the string ".55" equals the number 0.55. Would you like me to mimic excel, or make it so that in Lido this comparison would be true?

Yeah, spreadsheet data values have types. We should respect that too.

File: formulas/helpers.js [hunk 3]
```diff
@@ -299,6 +299,73 @@ class FormulaHelpers {
...
+                sum = sum * 100 + compare_value.charCodeAt(i) - lookup_value.charCodeAt(i)
+            }
+            let Longer_string = (lookup_value.length > compare_value.length) ? lookup_value : compare_value;
+            let direction = (lookup_value.length > compare_value.length) ? -1 : 1;
+            for(var i = min_Index; i < max_Index; i++){
+                sum = sum * 10 + direction * Longer_string.charCodeAt(i);
+            }
+            return sum;
+        }
+        //the special cases with the unique regex values
```
[the remark this answers is not in the document]
Comment (marhong2) [reply formulas/helpers.js:339]: Just to confirm, I should not implement ~ and instead use the wildcard-match library listed?

[the remark this answers is not in the document]
Comment (tyleradams) [reply formulas/helpers.js:339]: I was making a more subtle point, let's just ignore it so we can get this merged.

[the remark this answers is not in the document]
Comment (marhong2) [reply formulas/helpers.js:345]: FYI, EXCEL does not allow a user to assign match_mode to 2 and allow for BS. I was initially desigining arround the fact that I had to accommodate BS. Ill re-write it using the function you mentioned

**Pull request closed without merging**